master/internal/experiment.go

```diff
@@ -144,97 +151,45 @@ func newExperiment(master *Master, expModel *model.Experiment) (*experiment, err
-				return nil
-			}
-			seen[msg.Workload] = true
-
-			// Pass the workload completed message to the Trial. It will pass the event along to
-			// the experiment before this Ask gets a response.
-			master.system.AskAt(ref.Address().Child(requestIDs[msg.Workload.TrialID]), msg).Get()
-
-			// Wait for the experiment to handle any searcher operations due to the completed
-			// workload.
-			master.system.Ask(ref, doneProcessingSearcherOperations{}).Get()
-
-		case TrialClosedEventType:
-			// Ignore these events; the trial actors' closing will notify the experiment naturally.
-		}
-		return nil
-	}
+		faultToleranceEnabled: true,
+	}, nil
 }
 
+// Restore works by restoring from distributed consistent snapshots taken through the course
+// of an experiment. Snapshots within the system flow from the bottom up, starting with the
+// trial workload sequencer, to the trial, and finally to the experiment. Any event that the
+// trial or trial workload sequencer processes that would trigger a change to the state of the
+// experiment is:
+//   1. Propagated atomically, within a single message, to ensure the experiment handles it all
+//      or nothing
+//   2. With a snapshot affixed to it, to mark that it should trigger a snapshot
+// Upon receipt, the experiment handles the event entirely, snapshots its state and saves it
+// along with all the snapshots it has received, atomically.
+//
+// This is technically equivalent to the Chandy-Lamport distributed snapshotting algorithm where
+// the trial workload sequencer initiates the snapshots and propagates them along with trialCreated
+// and trialCompletedWorkload messages, with the snapshot itself acting as the marker. Each actor we
+// care about and know has updated state in the system snapshots it state when it receives this
+// marker and propagates it through the system. Even though all trials are part of each distributed
+// snapshot, we do not propagate snapshots to trials that didn't initiate the snapshot; that a trial
+// sends a snapshot with each state change implies if it did not send the snapshot, it does not have
+// a state change [1]. We can also be sure there are no pre-snapshot messages floating in the ether,
+// since any message we care about snapshotting is itself a snapshot.
+//
+// To restore properly, any message the experiment would send as a result of snapshot-able event
+// must be replayed on restore. In our case, these are the new operations created by the searcher
+// as a result a trialCreated or trialCompletedWorkload message. We restore trials and experiment
+// directly from their snapshots and tell the trial all its operations on creation.
+//
+// [1] This is fine, but _also_ we have to make sure that nothing else alters trial and experiment
+// state that doesn't flow from trial to experiment. For example, right now experiments update the
+// trials with what their state is. If we wanted to snapshot this, we would need to move to a full
```

Comment:
The task state ("active", "pending", whatever) should be a derived value at any time (only terminal states should be persisted, since they can't be altered).  The target state may differ based on what the user has requested, (pause, cancel, kill), but I don't see why those values would interact with the trial state snapshotting in any way; the operations requested is equal are unaffected, the checkpoints and validations taken are similarly unaffected.

(Actually, it doesn't look like you are even persisting the task state as part of the snapshot, so I'm extra confused about this comment)

Reply:
We talked about this offline, this was really just a hypothetical of something that would break our snapshotting scheme (break invariants that lead to invalid snapshots). I'll clarify this and remove the misleading example.